File: PracticeProblems/CPSolns/LeetCodeDSI/pathSum.cpp
#include<iostream>
#include "./treeStructure.cpp"
#include "./levelOrder.cpp"

using std::cout;
using std::endl;

/* 
 * PROBLEM STATEMENT:
 * Given root of binary tree and interger targetSum, return true if tree has
 * root-leaf path adding up to targetSum otherwise false;
 *
 * Approach:
 * Bases case if root is null
 * Traverse the left and right subtrees subtracting node values from target sum recursively
 * If leaf node is reached when target sum is 0, return true otherwise return false.
 */

class Solution {
    public:
        bool hasPathSum(TreeNode* root, int targetSum){
            if(!root)
                return false;
            else if(!root->left && !root->right && targetSum - root->data == 0)
                return true;
            else
                return hasPathSum(root->right, targetSum - root->data) || hasPathSum(root->left, targetSum - root->data);
        }
};

int main(){
    Solution soln;
    TraverseTree traverse;
    TreeNode* root = newNode(5);
    root->left = newNode(4);
    root->right = newNode(8);
    root->left->left = newNode(11);
    root->left->left->left = newNode(7);
    root->left->left->right = newNode(2);
    root->right = newNode(8);
    root->right->left = newNode(13);
    root->right->right = newNode(4);
    root->right->right->right = newNode(1);
    cout << soln.hasPathSum(root, 2) << endl;
    return 0;
}

File: Datastructures/PracticeProblems/LeetCodeDSI-JavaSolns/PathSum.java
// Give root of binary tree and target sum, return true if targetSum exists in tree 
// otherwise false
// Leet code 112

class PathSum{
    static TreeNode root;

    PathSum(int data){
        root = new TreeNode(data);
    }

    PathSum(){
        root = null;
    }

    static boolean hasPathSum(TreeNode root, int targetSum){
        if(root == null)
            return false;
        else if(root.left == null && root.right == null && targetSum - root.data == 0)
            return true;
        else
            return hasPathSum(root.right, targetSum - root.data) || hasPathSum(root.left, targetSum - root.data);
    }

    public static void main(String[] args){
        PathSum tree = new PathSum();
        tree.root = new TreeNode(5);
        root.left = new TreeNode(4);
        root.right = new TreeNode(8);
        root.left.left = new TreeNode(11);
        root.left.left.left = new TreeNode(7);
        root.left.left.right = new TreeNode(2);
        root.right.left = new TreeNode(13);
        root.right.right = new TreeNode(4);
        root.right.right.right = new TreeNode(1);
        System.out.println(hasPathSum(root, 22));
        System.out.println(hasPathSum(root, 89));
    }
}
